spidy/spidy.go

```diff
@@ -0,0 +1,381 @@
+
+var depths int64
+
+// collectFrom uses a recursive function to map out the needed lists of links to.
+// It returns a channel through which the acceptable links can be crawled from.
+func collectFrom(path *url.URL, doExternals bool, maxWorkers int, depths int, dead chan LinkReport, events Events) {
+	poolCfg := pool.Config{
+		OptEvent:    pool.OptEvent{Event: events.Event},
+		MinRoutines: func() int { return 10 },
+		MaxRoutines: func() int { return maxWorkers },
+	}
+
+	// create a new pool.
+	pl, err := pool.New("spidy", "collectFrom", poolCfg)
+	if err != nil {
+		fmt.Printf("Spidy failed to create work pool: %s\n", err.Error())
+		close(dead)
+		return
+	}
+
+	status, crawleable, err := evaluatePath(path.String())
+	if err != nil {
+		pl.Shutdown("spidy")
+		dead <- LinkReport{Link: path.String(), Status: status, Error: err}
+		close(dead)
+		return
+	}
+
+	defer pl.Shutdown("spidy")
+	defer close(dead)
+
+	if !crawleable {
+		return
+	}
+
+	var wait sync.WaitGroup
+	wait.Add(1)
+
+	// vl provides a rwmutex for control concurrent reads and writes on the visited
+	// map.
+	var vl sync.RWMutex
+
+	// visited is a map for storing visited uri's to avoid visit loops.
+	visited := make(map[string]bool)
+	visited[path.String()] = true
+
+	pl.Do("collectFrom", &pathBot{
+		path:      path.String(),
+		index:     path,
+		dead:      dead,
+		wait:      &wait,
+		vl:        &vl,
+		visited:   visited,
+		pool:      pl,
+		externals: doExternals,
+		skipCheck: true,
+		maxdepths: depths,
+	})
+
+	wait.Wait()
+
+	return
+}
+
+//==============================================================================
+
+// pathBot provides a worker which checks a giving URL path, cascading its
+// effects down its subroots and rescheduling new workers for those sublinks.
+// It implements pool.Work interface.
+type pathBot struct {
+	path      string
+	dead      chan LinkReport
+	wait      *sync.WaitGroup
+	vl        *sync.RWMutex
+	visited   map[string]bool
+	pool      *pool.Pool
+	index     *url.URL
+	skipCheck bool
+	externals bool
+	maxdepths int
+	cd        int
+}
+
+// Work performs the necessary tasks of validating a link and rescheduling
+// checks for sublinks.
+func (p *pathBot) Work(context interface{}, id int) {
+	// defer fmt.Printf("Ending task %d -> %s\n", id, p.path)
+	defer p.wait.Done()
+
+	p.vl.Lock()
+	p.visited[p.path] = true
+	p.vl.Unlock()
+
+	if !p.skipCheck {
+		status, crawleable, err := evaluatePath(p.path)
+		if err != nil {
+			p.dead <- LinkReport{Link: p.path, Status: status, Error: err}
+			return
+		}
+
+		if !crawleable {
+			return
+		}
+	}
+
+	links := make(chan string)
+
+	if err := farmLinks(p.path, links); err != nil {
+		fmt.Printf("Spidy Failed to Farm Links for Page[%s]: Error[%s]\n", p.path, err.Error())
+		p.dead <- LinkReport{Link: p.path, Status: http.StatusInternalServerError, Error: err}
+		return
+	}
+
+	for {
+		select {
+		case <-time.After(time.Second):
```

Comment:
What is the purpose of this timer if all it does is continue? It is not necessary. The case statement below will get triggered when something is received on the channel.